braindecode/models/attn_sleep.py

```diff
@@ -0,0 +1,395 @@
+            )
+
+        layers = []
+        layers.append(block(self.inplanes, planes, stride, downsample))
+        self.inplanes = planes * block.expansion
+        for i in range(1, blocks):
+            layers.append(block(self.inplanes, planes))
+
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+        x1 = self.features1(x)
+        x2 = self.features2(x)
+        x_concat = torch.cat((x1, x2), dim=2)
+        x_concat = self.dropout(x_concat)
+        x_concat = self.AFR(x_concat)
+        return x_concat
+
+
+##########################################################################################
+
+
+def attention(query, key, value, dropout=None):
+    """Implementation of Scaled dot product attention"""
+    d_k = query.size(-1)
+    scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(d_k)
+
+    p_attn = F.softmax(scores, dim=-1)
+    if dropout is not None:
+        p_attn = dropout(p_attn)
+    return torch.matmul(p_attn, value), p_attn
+
+
+class _CausalConv1d(torch.nn.Conv1d):
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size,
+                 stride=1,
+                 dilation=1,
+                 groups=1,
+                 bias=True):
+        self.__padding = (kernel_size - 1) * dilation
+
+        super(_CausalConv1d, self).__init__(
+            in_channels,
+            out_channels,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=self.__padding,
+            dilation=dilation,
+            groups=groups,
+            bias=bias)
+
+    def forward(self, input):
+        result = super(_CausalConv1d, self).forward(input)
+        if self.__padding != 0:
+            return result[:, :, :-self.__padding]
+        return result
+
+
+class _MultiHeadedAttention(nn.Module):
+    def __init__(self, h, d_model, afr_reduced_cnn_size, dropout=0.1):
+        """Take in model size and number of heads."""
+        super(_MultiHeadedAttention, self).__init__()
+        assert d_model % h == 0
+        self.d_k = d_model // h
+        self.h = h
+
+        self.convs = clones(_CausalConv1d(afr_reduced_cnn_size, afr_reduced_cnn_size, kernel_size=7, stride=1), 3)
+        self.linear = nn.Linear(d_model, d_model)
+        self.dropout = nn.Dropout(p=dropout)
+
+    def forward(self, query, key, value):
+        """Implements Multi-head attention"""
+        nbatches = query.size(0)
+
+        query = query.view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+        key = self.convs[1](key).view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+        value = self.convs[2](value).view(nbatches, -1, self.h, self.d_k).transpose(1, 2)
+
+        x, self.attn = attention(query, key, value, dropout=self.dropout)
+
+        x = x.transpose(1, 2).contiguous() \
+            .view(nbatches, -1, self.h * self.d_k)
+
+        return self.linear(x)
+
+
+class LayerNorm(nn.Module):
```

Comment:
why not using https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html ?